### PR TITLE
Tutorial on how to translate Script Lab

### PR DESCRIPTION
**Pull request by maninweb**
## Description
Tutorial on how to translate Script Lab to another language.

## Related Issue
#398

## Motivation and Context
Should help potential translators on how to translate Script Lab.

## How Has This Been Tested?
Not needed as it's a document with no code.

## Screenshots (if appropriate):
N/A

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: TRANSLATING.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+* [Understanding the Script Lab structure for translating](TRANSLATING.md#structure)
+* [Create a new translation](TRANSLATING.md/#create)
+* [Improve an existing translation](TRANSLATING.md#improve)
+* [Testing your translation](TRANSLATING.md#testing)
+* [Script Lab languages](TRANSLATING.md/#translations)
+
+<a id="prerequisites"></a>
+## Prerequisites
+
+Fork this project into your GitHub account and use branches to update existing files or add new files. When you are ready, send us a pull request. Please note, that if you are not employed by Microsoft and you have not already signed the Microsoft Contribution License Agreement, you will be asked to sign the agreement before your pull request is accepted.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: for "use branches", maybe be more specific?  I.e., "and create a branch for your translation work, which you will later use for a pull request"

File: TRANSLATING.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+We recommend you to download and install Visual Studio Code from here <https://code.visualstudio.com/> and follow the instructions from here [CONTRIBUTING.md](CONTRIBUTING.md) how to run the playground from source and having a local copy of the project.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: I think "how to run" should "*for* how to run"

File: TRANSLATING.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+We recommend you to download and install Visual Studio Code from here <https://code.visualstudio.com/> and follow the instructions from here [CONTRIBUTING.md](CONTRIBUTING.md) how to run the playground from source and having a local copy of the project. 
+
+<a id="structure"></a>
+## Understanding the Script Lab structure for translating
+
+The Script Lab code is mainly composed of two parts: the server part and the client part. Each part has its own files to be modified and/or added for creating or improving a translation. Additionally, the project includes 4 manifest files which also include translatable strings.
+
+Basically, the structure is represented by the following table. For the server and client part, each translation is stored in separate files named to the corresponding language. For each manifest file, the translations are included in different sections of the file.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: I think this table would be much easier to understand if it was actually three separate tables: server, client, and manifest.

File: TRANSLATING.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+const languageGenerator: { [key: string]: () => ServerStrings } = {
+    'en': () => getEnglishStrings(),
+    'de': () => getGermanStrings(),
+    'fr': () => getFrenchStrings(),
+    '??': () => createFakeStrings(() => getEnglishStrings())
+};
+```
+In the code above, the new line `import { getFrenchStrings } from './french';` has been added to the already existing imports for English and German. The second new code line is `'fr': () => getFrenchStrings(),` - which has been added to the language generator.
+
+That's all. Now you can start to translate your language file. Please note, that some strings include variables in their text. These variables look like `${Variable}` and should not be translated and also be preserved in your translation.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: Maybe clarify a bit:  "Please note that while most strings are just strings, others are functions that accept some parameters, and those parameters often get inlined into the return string using TypeScript template strings such as <add example here>".  For these variables, which look like `${parameterName}`, the name inside of `${parameterName}` should be preserved as is.

File: TRANSLATING.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    <Override Locale="fr" Value="Aide"/>
+  </bt:String>
+```
+
+<a id="improve"></a>
+## Improve an existing translation
+
+If you would like to improve an existing translation, e.g. correct a typing error or suggest a better wording, then just open the corresponding existing language files and do your changes. Please note, that when modifying one of the manifest files, you should also do the change in the 3 other files.
+
+For example, if you discovered an error in the translation to *German* for the *client part*, then open the file `german.ts` from the folder `src/client/app/strings` and do the change.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: I would move that sentence up by one, and the manifest sentence down by own to a separate paragraph.  Otherwise I mistakenly thought that this sentence is also about the manifest.

File: TRANSLATING.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+<a id="testing"></a>
+## Testing your translation
+
+You should test your translations locally before pulling it to the original repository. Testing helps you to validate your translation and see how it looks in the application. And, for example, see if texts are too long and should be shortened. For creating a testing environment, please refer to the instructions from here [CONTRIBUTING.md](CONTRIBUTING.md).
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: "before pulling it" ==> I would rephrase it to "before sending a pull request to the original repository"

File: TRANSLATING.md (hunk 1 above)
Comment (maninweb) [inline TRANSLATING.md]: Done, I have added links to the GitHub Help too.

File: TRANSLATING.md (hunk 2 above)
Comment (maninweb) [inline TRANSLATING.md]: Done, thanks :-)

File: TRANSLATING.md (hunk 3 above)
Comment (maninweb) [inline TRANSLATING.md]: Changed - and yes, looks much prettier now :-)

File: TRANSLATING.md (hunk 4 above)
Comment (maninweb) [inline TRANSLATING.md]: Done.

File: TRANSLATING.md (hunk 5 above)
Comment (maninweb) [inline TRANSLATING.md]: Done.

File: TRANSLATING.md (hunk 6 above)
Comment (maninweb) [inline TRANSLATING.md]: Yes, much better, done :-)

**Comment by maninweb** [comment]
I have done some more changes too:

* Changed the link for the language country codes to Office.
I referred for this on the link from Office JS Docs https://github.com/OfficeDev/office-js-docs/blob/master/docs/develop/localization.md but updated it to Office 2016. May be, Office JS Docs should update their link too (on the bottom Additional Resources)?
* Added text for special characters in the ts-files and xml-files.
For the ts-files, I added a paragraph how to escape the single quote (the french friend have a lot of them)
For the XML-files, I added a text too. However I am not sure, if this is totally correct. In the Office JS doc for localization, it is not mentionned. May be a suggestion to add some word about special characters there?
* Corrected some typing/grammatical errors.

Best,
Mourad

File: TRANSLATING.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    'de': () => getGermanStrings(),
+    'fr': () => getFrenchStrings(),
+    '??': () => createFakeStrings(() => getEnglishStrings())
+};
+```
+In the code above, the new line `import { getFrenchStrings } from './french';` has been added to the already existing imports for English and German. And the second new code line `'fr': () => getFrenchStrings(),` has been added to the language generator.
+
+That's all. Now you can start to translate your language file. Please note, that while most strings are just strings, others are functions that accept some parameters, and those parameters often get inlined into the return string using TypeScript template strings such as double quotes `"`. For these variables, which look like `${parameterName}`, the name inside of `${parameterName}` should be preserved as is.
+
+The strings in the TypeScripe files are enclosed by single quotes or apostrophes `'`. If you need to use the single quote character inside a string, you need to escape this character by prepending a backslash. Example: the word `today` then translates in french to `aujourd\'hui`.
```
Comment (Zlatkovsky) [inline TRANSLATING.md]: Actually (both FYI, and something to put in here), the simplest way to avoid the escaping problem is to use the TypeScript template strings, even if you don't do any variables within them!

so use the ` symbol, as so:

    something: `It's a wonderful "thing" that you don't have to escape anything anymore!`

**Comment by Zlatkovsky** [comment]
@maninweb , I added one more comment re. templated strings (and the need to escape 's and "s).  I think it will help you to know it in the future, and will help other translators as well.  Could this make this one final change?

For your other comments:  I am not sure I understood what you meant by "Maybe, Office JS Docs updates their link too (on the bottom Additional Resources)?".  Though you can always submit a pull request to their docs too, with a suggestion for the change :-).  In all of my experiences with the Docs team, they are delighted to receive contributions.

For escaping XML, looks like it's just the three characters of &, ", and ' that need escaping.  But again, feel free to submit a pull request to the docs, if you think it should be mentioned there.

Finally, you may have received a link from GitHub adding you to the Script Lab repository.  It looks like you need to be a repo member in order for us to be able to assign issues (like the translation bug) to you, once an issue has been closed.  So... welcome to the repo :-)

**Comment by maninweb** [comment]
@Zlatkovsky, thanks for the invite to the repo :-) I changed the escaping text in the document. Regarding my comments for the localization doc on Office JS, yes, thanks, I will post and issue/suggestion/pull request there :-)
